app/hooks/useDownload.ts

```diff
@@ -0,0 +1,45 @@
+import { useEffect, useRef, useState } from "react"
+import { DownloadState, trackDownloadsClient } from "../clients/TrackDownloadsClient"
+import { Track } from "../models"
+
+type UseDownloadResponse = {
+  data: DownloadState
+  start: () => void
+  clear: () => void
+  pause: () => void
+}
+
+export function useDownload(track: Track): UseDownloadResponse {
+  const isMountedRef = useRef(false)
+  const [, forceRerender] = useState(0)
+
+  useEffect(() => {
+    isMountedRef.current = true
+    const unsubscribe = trackDownloadsClient.subscribe(track.trackId, () => {
+      if (isMountedRef.current) {
+        forceRerender((x) => x + 1)
+      }
+    })
+    trackDownloadsClient.findDownloadInLocalFileSystem(track)
+
+    return () => {
+      isMountedRef.current = false
+      unsubscribe()
+    }
+  }, [])
```

Comment:
Why is the array empty and does not include `isMountedRef` and `forceRerender`? Somewhere in the React documentation I read that you must list all dependencies because otherwise unexpected things may happen.

Reply:
Yeah, you are right. I just wanted to blindly adhere to the [`exhaustive-deps`](https://github.com/facebook/react/issues/14920) rule because it works even without in-depth knowledge about references and state setters. The note in https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects also says

> there are usually better solutions to avoid re-running effects too often

and refer to [Is it safe to omit functions from the list of dependencies?](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies) which provides various refactorings.

I would leave it as is, I just wanted to point it out. I stumbled upon this once at work because the linter produced many warnings when I wanted to do the same thing: Just have an effect that runs exactly once.

I just remember that I also had a problem with this when using `audio-player` and a component parameter changed that did not remount the component but just re-rendered it and I had a `useEffect` with `[]` that depended on that parameter though and my audio did not change.